---
title: "Intro to Microbiome"
author: "Jim Arnold"
date: "August 10, 2018"
output: html_document
---

Intro to microbiome analysis in R using the R markdown format so it'll be:
a) interactive
b) notated so I can remember what I was thinking! 
Good looking out past-self, let's do this!

I found a neat package on github and I want to work through some of their examples. Link: https://microbiome.github.io/microbiome/Tutorial.html

First, install the microbiome package. If it's a fresh install, you need to get it from bioconductor.

```{r}
library(microbiome)
library(knitr)
```

Load in the data, 
>This data set from Lahti et al. Nat. Comm. 5:4344, 2014 comes with 130 genus-like taxonomic groups across 1006 western adults with no reported health complications. Some subjects have also short time series. 

```{r}
data(atlas1006) 
pseq <- atlas1006
print(pseq)
```
This is phyloseq object. There are special calls for it, listed here:
https://microbiome.github.io/microbiome/Preprocessing.html

A phyloseq object contains OTU table (taxa abundances), sample metadata, taxonomy table (mapping between OTUs and higher-level taxonomic classifications), and phylogenetic tree (relations between the taxa). Some of these are optional.

```{r}
summarize_phyloseq(pseq)
```

A phyloseq object contains OTU table (taxa abundances), sample metadata, taxonomy table (mapping between OTUs and higher-level taxonomic classifications), and phylogenetic tree (relations between the taxa). Some of these are optional.

Pull the relevant data elements from the phyloseq object:
metadata
taxonomy
abundance

```{r}
meta <- meta(pseq)
head(meta)
```
```{r}
taxonomy <- tax_table(pseq)
head(taxonomy)
```
```{r}
#absolute abundances
otu_absolute <- abundances(pseq)

#relative abundances
otu.relative <- abundances(pseq, "compositional")
```

Recommended: melt the phyloseq data for easier plotting:
```{r}
df <- psmelt(pseq)
head(df)
```

Data can be accessed by sample metadata and conditionals:
```{r}
#Abundance of a given species in each sample
head(abundances(pseq)["Akkermansia",])

#Select a subset by metadata fields:
pseq.subset <- subset_samples(pseq, nationality == "US")

# Check sample names for African Females in this phyloseq object
s <- rownames(subset(meta(pseq), nationality == "US" & gender == "male"))
# Pick the phyloseq subset with these sample names
pseq.subset2 <- prune_samples(s, pseq)

```

Explore the estimation and analysis of various diversity indices and taxonomic composition.
Diversity - https://microbiome.github.io/microbiome/Diversity.html
taxonomic comp - https://microbiome.github.io/microbiome/Composition.html
```{r}
tab <- global(pseq, index = "all")
head(tab)
```

Explore potential technical biases in the data. DNA extraction method reportedly has a large effect on sample grouping.
```{r}
# Use relative abundance data
ps <- microbiome::transform(pseq, "compositional")

# For this example, choose samples with DNA extraction information available
ps <- subset_samples(ps, !is.na(DNA_extraction_method))

# Or: you could focus on a single DNA extraction method
# ps <- subset_samples(ps, DNA_extraction_method == "r")
```

Plot PCoA by DNA extraction method.
Note - this step takes a while.

```{r}
# Illustrate sample similarities with PCoA (NMDS)
plot_landscape(ps, "NMDS", "bray", col = "DNA_extraction_method")
```

Clear difference between 'r' and 'o' DNA extraction methods. 'p' looks interspersed with 'o'.


Core microbiota - https://microbiome.github.io/microbiome/Core.html

A full phyloseq object with just the core taxa is obtained as follows:
```{r}
# Transform to compositional abundances
pseq.rel <- microbiome::transform(pseq, "compositional")

# Pick the core (>0.1% relative abundance in 10% of the samples)
pseq.core <- core(pseq.rel, detection = 0.1/100, prevalence = 10/100)

#Visualize
prevalences <- seq(.05, 1, .05)
detections <- 10^seq(log10(1e-3), log10(.2), length = 10)

p <- plot_core(pseq.core, plot.type = "heatmap",
    prevalences = prevalences, detections = detections) +
    xlab("Detection Threshold (Relative Abundance)")
print(p) 
```
Super interesting, you can play with the variables and id different taxa on the basis of frequency and/or abundance. 





